### PR TITLE
Update: Add "consistent" option to array-element-newline (fixes #9457)

### LINKED ISSUE #9457: Proposal: Add "consistent" option to array-element-newline
**Tell us about your environment**

* **ESLint Version:** 4.9.0
* **Node Version:** 8.7.0
* **npm Version:** 5.4.2

I would like to propose to add a third, new option, `consistent`, to the rule [array-element-newline](https://eslint.org/docs/rules/array-element-newline).

### Expected behaviour

Given `'array-element-newline': ['warn', 'consistent']`, a newline would be required between all array elements if any of the elements are placed on a separate line.

#### Good

```js
// 'array-element-newline': ['warn', 'consistent']
const arr = [1,2,3]
const arr = [
  1, 2, 3
]
const arr = [
  1,
  2,
  3,
]
```

#### Bad

```js
// 'array-element-newline': ['warn', 'consistent']
const arr = [
  1, 2,
  3,
]
```

#### Autofix

If implemented, the autofix would separate all elements onto a separate line if any of the elements are on a separate line. The autofix would never put all elements into a single line.

### Current behaviour

The rule currently only supports only:

- no newlines between elements
- `num` elements-based decision about elements placement
- all elements on their own line

-----

This follows the idea recently implemented in #9206 for `array-bracket-newline` and is tangentially related to #9054.

### PR DESCRIPTION
**Pull request by g-plane**
close #9457

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

https://github.com/eslint/eslint/issues/9457

**What changes did you make? (Give an overview)**
Add a new option `"consistent"` for rule `array-element-newline`.

**Is there anything you'd like reviewers to focus on?**

### CODE REVIEW

File: tests/lib/rules/array-element-newline.js [hunk 1]
```diff
@@ -66,6 +66,23 @@ ruleTester.run("array-element-newline", rule, {
         { code: "var foo = [1, (\n2\n), 3];", options: ["never"] },
         { code: "var foo = [\nfunction foo() {\ndosomething();\n}, function bar() {\ndosomething();\n}\n];", options: ["never"] },
 
+        // "consistent"
+        { code: "var foo = [];", options: ["consistent"] },
+        { code: "var foo = [1];", options: ["consistent"] },
+        { code: "var foo = [1, 2];", options: ["consistent"] },
+        { code: "var foo = [1,\n2];", options: ["consistent"] },
```
Comment (ljharb) [inline tests/lib/rules/array-element-newline.js]: hm, why is this consistent? I'd expect "consistent" to require either "every item is on the same line, along with the brackets" or "every item is on its own line, as is each bracket" - but nothing in between.

Comment (not-an-aardvark) [reply tests/lib/rules/array-element-newline.js]: I think brackets are handled separately with [`array-bracket-newline`](https://eslint.org/docs/rules/array-bracket-newline). `array-element-newline` just deals with elements and not brackets.

Comment (g-plane) [reply tests/lib/rules/array-element-newline.js]: This rule should not concern the style of brackets.

Comment (ljharb) [reply tests/lib/rules/array-element-newline.js]: hm, ok - i suppose with "multiline", array-bracket-newline would work in concert with this rule and "consistent" to do what I'd expect.

**Comment by g-plane** [comment]
@platinumazure Just according to #9457 :
> If implemented, the autofix would separate all elements onto a separate line if any of the elements are on a separate line. The autofix would never put all elements into a single line.

**Comment by g-plane** [comment]
Updated.

**Comment by platinumazure** [comment]
This has sat long enough so I'm going to go ahead and merge. Thanks @g-plane for contributing!

**Pull request merged**